internal/celoapi/backend.go

```diff
@@ -7,91 +7,95 @@ import (
+	if ok {
+		return cachedRates, nil
+	}
+	cb, err := b.getContractCaller(ctx, atBlock)
 	if err != nil {
 		return nil, err
 	}
 	er, err := contracts.GetExchangeRates(cb)
 	if err != nil {
 		return nil, fmt.Errorf("retrieve exchange rates from current state: %w", err)
 	}
-	return exchange.ConvertGoldToCurrency(er, feeCurrency, nativePrice)
+	b.exchangeRatesCache.Add(atBlock, er)
+	return er, nil
 }
 
-// GasPrice wraps the original JSON RPC `eth_gasPrice` and adds an additional
-// optional parameter `feeCurrency` for fee-currency conversion.
-// When `feeCurrency` is not given, then the original JSON RPC method is called without conversion.
-func (c *CeloAPI) GasPrice(ctx context.Context, feeCurrency *common.Address) (*hexutil.Big, error) {
-	tipcap, err := c.ethAPI.GasPrice(ctx)
+func (b *CeloAPIBackend) ConvertToCurrency(ctx context.Context, atBlock common.Hash, value *big.Int, fromFeeCurrency *common.Address) (*big.Int, error) {
+	er, err := b.GetExchangeRates(ctx, atBlock)
 	if err != nil {
 		return nil, err
 	}
-	// Between the call to `ethapi.GasPrice` and the call to fetch and convert the rates,
-	// there is a chance of a state-change. This means that gas-price suggestion is calculated
-	// based on state of block x, while the currency conversion could be calculated based on block
-	// x+1.
-	// However, a similar race condition is present in the `ethapi.GasPrice` method itself.
-	return c.convertedCurrencyValue(tipcap, feeCurrency)
+	currencyValue, err := exchange.ConvertGoldToCurrency(er, fromFeeCurrency, value)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert to native from fee currency (%s): %w ", fromFeeCurrency, err)
+	}
+	return currencyValue, nil
 }
 
-// MaxPriorityFeePerGas wraps the original JSON RPC `eth_maxPriorityFeePerGas` and adds an additional
-// optional parameter `feeCurrency` for fee-currency conversion.
-// When `feeCurrency` is not given, then the original JSON RPC method is called without conversion.
-func (c *CeloAPI) MaxPriorityFeePerGas(ctx context.Context, feeCurrency *common.Address) (*hexutil.Big, error) {
-	tipcap, err := c.ethAPI.MaxPriorityFeePerGas(ctx)
+func (b *CeloAPIBackend) ConvertToGold(ctx context.Context, atBlock common.Hash, value *big.Int, toFeeCurrency *common.Address) (*big.Int, error) {
+	er, err := b.GetExchangeRates(ctx, atBlock)
 	if err != nil {
 		return nil, err
 	}
-	// Between the call to `ethapi.MaxPriorityFeePerGas` and the call to fetch and convert the rates,
-	// there is a chance of a state-change. This means that gas-price suggestion is calculated
-	// based on state of block x, while the currency conversion could be calculated based on block
-	// x+1.
-	return c.convertedCurrencyValue(tipcap, feeCurrency)
+	goldValue, err := exchange.ConvertCurrencyToGold(er, value, toFeeCurrency)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert from native to fee currency (%s): %w ", toFeeCurrency, err)
+	}
```

Comment:
As above, this error message sounds like it would fit for all calls to `exchange.ConvertCurrencyToGold` and should therefore be inside `exchange.ConvertCurrencyToGold` rather than outside of if.